### ISSUE TITLE
Support foreign key constraints for colocated tables

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#698]
Once we formally introduce the notion of table colocation (#695), we can then add support for foreign key constraints on colocated tables. (A foreign key constraint specifies that the values in a column (or a group of columns) must match the values appearing in some row of another table. We say this maintains the referential integrity between two related tables.)

A simple way to do this is to deparse the foreign key constrains, check table colocation, and pass them down to related shards.

**Comment by onderkalaci** [comment]
Note that @byucesoy shared earlier on:

**Foreign Key Constraints Scope**

Given that tables are co-located and foreign constraint is defined on partition columns of particular tables, we can support almost all foreign key related functionality and provide referential integrity. This document defines scope of foreign key constraint support for Citus 5.3 (or 6.0)

**Requirements**

* Tables should be co-located.
* Referenced column or column group should be unique or primary key. (Postgresql enforces this already)
* Both table’s partition columns should be part of column list in respective side of foreign key clause
* If foreign key references group of columns, order of partition column should be same place in both groups.

**What We Will Support**
* We will support foreign keys referencing single column or group of columns.
* We will support tables having more than one foreign key constraint.
* We will support ON DELETE action and ON UPDATE action keywords when
action is NO ACTION, RESTRICT or CASCADE.
* We will support DEFERRABLE, NOT DEFERRABLE, INITIALLY DEFERRED and INITIALLY IMMEDIATE keywords.
* We will support MATCH SIMPLE and MATCH FULL keywords. (MATCH PARTIAL is not implemented in PostgreSQL yet)
* We will support creating foreign keys both as column constraint and table constraint. This is just syntactic issue and will be handled by PostgreSQL’s parser. (See CREATE TABLE documentation page for column constraint and table constraint syntax)
* We will support ALTER TABLE ADD CONSTRAINT depending on schedule.

**What We Will Not Support**
* We will not support defining foreign key constraint on arbitrary columns
* We will not support ON DELETE action and ON UPDATE action keywords when
action is SET NULL or SET DEFAULT, because we do not want to set partition column to NULL or a default value.
* Some Examples of What We Will Support

For now, we create co-located tables by setting shard count and replication factor of tables to same number, but with 5.3 release we will have well-defined co-location API.

* Foreign key constraint from partition column to partition column;

```SQL
CREATE TABLE products (
    product_id integer PRIMARY KEY,
    ...
);
SELECT master_create_distributed_table('products', 'product_id, 'hash');
SELECT master_create_worker_shards('products', 16, 1);


CREATE TABLE orders (
    order_id integer,
    product_id integer REFERENCES products (product_id),
    ...
);
SELECT master_create_distributed_table('orders', 'product_id, 'hash');
SELECT master_create_worker_shards('orders', 16, 1);

```

*  Foreign key constraint referencing group of columns & table having multiple foreign constraints;

```SQL
CREATE TABLE products (
    product_id integer PRIMARY KEY,
    ...
);
SELECT master_create_distributed_table('products', 'product_id', 'hash');
SELECT master_create_worker_shards('products', 16, 2);



CREATE TABLE orders (
    product_id integer,
    order_id integer,
    ...
    PRIMARY KEY (product_id, order_id)
);
SELECT master_create_distributed_table('orders', 'product_id', 'hash');
SELECT master_create_worker_shards('orders', 16, 2);



CREATE TABLE order_items (
    order_item_id integer PRIMARY KEY,
    product_id integer,
    order_id integer,
    ...
    FOREIGN KEY product_id REFERENCES products (product_id),
    FOREIGN KEY (product_id, order_id) REFERENCES orders (product_id, order_id)
);
SELECT master_create_distributed_table('order_items', 'product_id', 'hash');
SELECT master_create_worker_shards('order_items', 16, 2);
```


Some more implementation details:

In summary, we could separate GetTableDDLEvents() and GetTableForeignConstraints(). In return, we could only use GetTableForeignConstraints() whenever necessary (literally create_worker_shards and shard rebalancer).

* psudo-code for worker shard creation:

```C
create_woker_shards()
{
ddlEvents = getTableDDLEvents()

foreach placement
    worker_apply_ddl_commands(shardId, ddlCommand)

###### start foreign key part
foreignConstraints = getTableForeignConstraints()

foreach placement
    referencedShardId = FindReferenceShardId(shardId, foreignConstraint)
    worker_apply_foreign_constraints(shardId, referencedShardId, foreignConstraint)

}
```
**Rebalance operation would be very similar:**
* Drop all colocated placements on the target with a single command
* GetTableDDLEvents for all placements (i.e. apply worker_apply_ddl_commands) + Fetch Data
* GetTableForeignConstraints() for all collocated tables, then, worker_apply_foreign_constraints for each constraint
* Drop all colocated placements on the source with a single command

In this way, we never touch how real-time/task-tracker executors run. We only create foreign constraints when they are actually required. We also will not need to change signature of worker_apply_ddl_commands which will save us upgrading problems during rolling upgrades.

From the time line perspective, we could implement foreign keys for worker_create_shards() in 1 week. We could implement the rebalancer part while we're implementing the rebalancer. It'd probably become much simpler given that we already implemented infrastructure for creating the foreign constraints.

(From the design perspective, having a separate constructs for foreign keys makes sense given that all the other DDL events are only related to a single table. In this case, we're dealing with multiple tables at a time.)

**Comment by ozgune** [comment]
@byucesoy -- Now that we have foreign keys implemented in v6.0, are we good to close this issue?

**Comment by byucesoy** [comment]
We only have foreign keys if they are created in CREATE TABLE statements. We do not push down foreign key creation for ALTER TABLE queries. We have open PR for it #876 but it needs some heavy refactoring at the moment. We can either:
- Close this issue when we merge #876 
- OR Close this issue and add another issue for supporting foreign keys with ALTER TABLE queries.
